### PR TITLE
Upgrade to TS 2.1

### LINKED ISSUE #11: Object.assign missing for IE11
screen shot 2016-10-26 at 10 22 04 pm

### LINKED ISSUE #376: Upgrade to TypeScript 2.1
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html

- [x] switch from Object.assign to ES2017 object spread
  - also fixes #11 
- [ ] configuration inheritance (likely useful for the `examples/` TS project)
  - :sob: not supported yet https://github.com/ivogabe/gulp-typescript/issues/459
- [ ] `Partial<T>` (haven't looked into this much yet for this codebase)

these should all be non-breaking changes.

### PR DESCRIPTION
**Pull request by giladgray**
#### Checklist

- [x] `tslib`
- [x] `Partial` for props in some test suites
- [x] `extends` outside of gulp-typescript 😭  https://github.com/ivogabe/gulp-typescript/issues/459

#### Changes proposed in this pull request:

- update typescript and gulp-typescript
- replace `Object.assign` and `Utils.shallowClone` with object spreads
- some lint fixes in table examples
- ⚠️  new dependency on `tslib` for `--importHelpers` flag
  - this reduces the size of each source file at the cost of one additional NPM dependency
- use `Partial<Props>` in test suites for components that have required props!
- use `extends` in tsconfig.json files that are never built with `gulp-typescript`

### CODE REVIEW

File: packages/core/src/common/utils.ts [hunk 1]
```diff
@@ -50,17 +50,6 @@ export function clamp(val: number, min: number, max: number) {
     return Math.min(Math.max(val, min), max);
 }
 
-/** Return a new object with the same keys as the given object (values are copied, not cloned). */
-export function shallowClone<T>(object: T): T {
```
Comment (adidahiya) [inline packages/core/src/common/utils.ts]: nice

File: packages/table/preview/index.tsx [hunk 2]
```diff
@@ -40,51 +40,52 @@ import {
     Utils,
 } from "../src/index";
 
+// tslint:disable:max-classes-per-file no-console jsx-no-lambda jsx-no-multiline-js
```
Comment (adidahiya) [inline packages/table/preview/index.tsx]: do we _need_ to disable jsx-no-multiline-js? would prefer to keep it

File: packages/table/preview/index.tsx [hunk 3]
```diff
@@ -377,11 +378,11 @@ ReactDOM.render(
...
     }, {
-        isColumnResizable: false,
-        isRowResizable: false,
-        selectionModes: SelectionModes.NONE,
-    }),
-    document.getElementById("table-1")
+            isColumnResizable: false,
+            isRowResizable: false,
+            selectionModes: SelectionModes.NONE,
+        }),
```
Comment (adidahiya) [inline packages/table/preview/index.tsx]: this indentation change looks wrong, the previous alignment is preferred

File: packages/table/preview/index.tsx [hunk 4]
```diff
@@ -443,18 +444,18 @@ ReactDOM.render(
...
+                {
+                    className: "my-group",
+                    regions: [
+                        Regions.cell(0, 0),
+                        Regions.row(2),
+                        Regions.cell(1, 2, 5, 2),
+                    ],
+                },
+            ],
+        }),
```
Comment (adidahiya) [inline packages/table/preview/index.tsx]: same here

**Comment by adidahiya** [comment]
what do you plan to do with tslib and Partial?

File: packages/table/preview/index.tsx (hunk 2 above)
Comment (giladgray) [inline packages/table/preview/index.tsx]: same but would prefer not to touch this file too much. most of what you see here was done automatically by TSLint or VSCode formattin.

**Comment by giladgray** [comment]
implement them :) marking this WIP for now

**Comment by giladgray** [comment]
FWIW, i'm able to use `extends` in a few places, like test tsconfigs, which are only ran via webpack. might be able to do examples too for the same reason. 

so it really only affects `packages/*/tsconfig.json`, the ones that get run through gulp-typescript.

File: packages/table/preview/index.tsx (hunk 3 above)
Comment (giladgray) [inline packages/table/preview/index.tsx]: sure, i didn't look closely at the formatting. maybe i'll just revert, it's not super relevant.

**Comment by giladgray** [comment]
@themadcreator @gscshoyru Table code coverage is failing the build. seems like a test for column resizing would fix it swimmingly. might one of you be able to take that one this week?

https://1214-71939872-gh.circle-artifacts.com/0/home/ubuntu/blueprint/packages/table/coverage/PhantomJS%202.1.1%20(Linux%200.0.0)/src/table.tsx.html

**Comment by adidahiya** [comment]
@giladgray why would coverage change as a result of this PR? Those things should really be decoupled; please don't add a test in this PR.

**Comment by giladgray** [comment]
i'm not planning to add a test in this PR and I'm not sure why coverage changed. you'll notice it's _extremely_ close to the threshold (79%, must be 80% to pass) so all it would have taken is one little tweak to drop coverage below the threshold. my guess is that switching to `tslib` removed enough code from the file as to reveal this.

**Comment by giladgray** [comment]
ready 4 review 👀

**Comment by adidahiya** [comment]
nvm, I read the [2.1 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html), this looks legit

**Pull request merged**